controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -914,9 +943,62 @@ func hostHasFinalizer(host *metal3v1alpha1.BareMetalHost) bool {
 	return utils.StringInList(host.Finalizers, metal3v1alpha1.BareMetalHostFinalizer)
 }
 
+// Resync the set at the startup
+func (r *BareMetalHostReconciler) syncProvisioningHosts(mgr ctrl.Manager) {
+	hosts := &metal3v1alpha1.BareMetalHostList{}
+	err := mgr.GetAPIReader().List(context.TODO(), hosts)
+	if err != nil {
+		os.Exit(1)
+	}
+
+	for _, host := range hosts.Items {
+		r.checkProvisioningHost(host, host.Status.Provisioning.State)
+	}
+}
+
+// Check if there's a free slot for hosts to be provisioned
+func (r *BareMetalHostReconciler) checkProvisioningHost(host metal3v1alpha1.BareMetalHost, state metal3v1alpha1.ProvisioningState) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	switch state {
+	case metal3v1alpha1.StateInspecting, metal3v1alpha1.StateProvisioning:
+		if len(r.currentProvisioningHosts) >= maxProvisioningHosts {
+			return false
+		}
+		r.currentProvisioningHosts[host.Name] = struct{}{}
+	default:
+		delete(r.currentProvisioningHosts, host.Name)
+	}
+
+	return true
+}
+
+// Check if there's a free slot for hosts that have been previously delayed
+func (r *BareMetalHostReconciler) checkDelayedHost(info *reconcileInfo) (bool, actionResult) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	switch info.host.Status.ErrorType {
+	case metal3v1alpha1.TooManyHostsError:
+		if len(r.currentProvisioningHosts) >= maxProvisioningHosts {
+			// No available slots, current action delayed
+			return true, recordActionFailure(info, metal3v1alpha1.TooManyHostsError, "Delayed, too many hosts")
+		}
+
+		// A slot could be available, let's cleanup the host and retry
+		info.host.ClearError()
+		return true, actionContinue{}
+	}
+
+	return false, nil
+}
+
 // SetupWithManager reigsters the reconciler to be run by the manager
 func (r *BareMetalHostReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
+	r.syncProvisioningHosts(mgr)
```

Comment:
I believe there is a race here; we haven't yet started watching resources or obtained the leader lock, so Hosts can be modified after we build our initial state without us seeing the changes (potentially ever, since our updates are edge-triggered).